src/Microsoft.NetCore.Analyzers/Core/Security/UseAutoValidateAntiforgeryToken.cs

```diff
@@ -0,0 +1,276 @@
+                            .Union(methodSymbol.TypeArguments);
+
+                        foreach (var potentialAntiForgeryFilter in potentialAntiForgeryFilters)
+                        {
+                            if (potentialAntiForgeryFilter.AllInterfaces.Contains(iFilterMetadataTypeSymbol) &&
+                                s_AntiForgeryRegex.IsMatch(potentialAntiForgeryFilter.Name))
+                            {
+                                hasGlobalAntiForgeryFilter = true;
+
+                                break;
+                            }
+                            else if (potentialAntiForgeryFilter.AllInterfaces.Contains(iAsyncAuthorizationFilterTypeSymbol))
+                            {
+                                onAuthorizationAsyncMethodSymbols.Add(
+                                    potentialAntiForgeryFilter
+                                    .GetMembers()
+                                    .OfType<IMethodSymbol>()
+                                    .FirstOrDefault(
+                                        s => s.Name == "OnAuthorizationAsync" &&
+                                            s.ReturnType.Equals(taskTypeSymbol) &&
+                                            s.Parameters.Length == 1 &&
+                                            s.Parameters[0].Type.Equals(authorizationFilterContextTypeSymbol)));
+                            }
+                        }
+                    }
+                }, OperationKind.Invocation);
+
+                compilationStartAnalysisContext.RegisterSymbolAction(symbolAnalysisContext =>
+                {
+                    var controllerTypeSymbol = (INamedTypeSymbol)symbolAnalysisContext.Symbol;
+
+                    // The controller class is protected by a validate anti forgery token attribute
+                    if (controllerTypeSymbol.GetAttributes().Any(s => s_AntiForgeryAttributeRegex.IsMatch(s.AttributeClass.Name)))
+                    {
+                        usingValidateAntiForgeryAttribute = true;
+
+                        return;
+                    }
+
+                    foreach (var actionMethodSymbol in controllerTypeSymbol.GetMembers().OfType<IMethodSymbol>())
+                    {
+
+                        // The method is protected by a validate anti forgery token attribute
+                        if (actionMethodSymbol.GetAttributes().Any(s => s_AntiForgeryAttributeRegex.IsMatch(s.AttributeClass.Name)))
+                        {
+                            usingValidateAntiForgeryAttribute = true;
+
+                            return;
+                        }
+
+                        var baseTypes = controllerTypeSymbol.GetBaseTypes();
+
+                        if ((baseTypes.Contains(controllerTypeSymbol) ||
+                            baseTypes.Contains(controllerBaseTypeSymbol)) && // An subtype of `Microsoft.AspNetCore.Mvc.Controller` or `Microsoft.AspNetCore.Mvc.ControllerBase`
```

Comment:
The first thing we should do in the SymbolAction callback is make sure this is an ASP.NET Core MVC controller, rather than something else.

Reply:
We just need to figure out if validate anti forgery token attributes is used somewhere to mitigate CSRF issues. Why it has to be an ASP.NET Core MVC controller.

Reply:
You want to ignore classes that aren't an ASP.NET Core MVC controller, since this rule is for ASP.NET Core MVC, right?  You don't want to set usingValidateAntiForgeryAttribute to true if it's not an ASP.NET Core MVC controller.

Reply:
Just checking, the case below can be marked with NoDiagnostic?

using Microsoft.AspNetCore.Mvc;

[MyValidateAntiForgeryAttribute]
class MakeSureValidateAntiForgeryAttributeIsUsedSomeWhereClass
{
}

class TestClass : ControllerBase
{
    [HttpPost]
    public override AcceptedAtActionResult AcceptedAtAction (string actionName)
    {
        return null;
    }
}

Reply:
Correct.